### DISCUSSION TITLE
Measuring kernel socket buffers in shadow (particularly wrt KIST)

### DISCUSSION THREAD

**Discussion by sporksmith** [shadow/shadow#1865] [General]
Summarizing some IRC/slack discussion for posterity and potential follow-up.

@mikeperry-tor observed that "switching all relays to kistlite actually reduced my cell queues with congestion control. and switching to vanilla scheduler pretty much eliminated them entirely", but now is trying to figure out if it's just pushing the buffers from tor userspace into the kernel.

@robgjansen  confirmed that:

> The old scheduler (which I think is what you call vanilla) just pushes as much to the kernel as the kernel will allow and Tor’s queues will indeed be lower or even empty
> This might be OK for normal apps, but for apps like Tor where relays have hundreds or even thousands of open connections, the data will just end up sitting in the kernel for a long time
> This is bad, because once Tor writes it to the kernel it has no more control over traffic priority, and things can change a lot from when tor writes it the first time until when the kernel decides to send it
> The whole point of KIST is to keep cells inside of Tor as long as possible, until we think the kernel is actually ready to send

More about KIST and this issue in particular:
* KIST conference paper and slides: https://www.usenix.org/conference/usenixsecurity14/technical-sessions/presentation/jansen
* KIST more-complete journal paper: https://www.robgjansen.com/publications/kist-tops2018.pdf

If we want to log more in Shadow:

> The total input buffer length is
`socket_getInputBufferLength(&(tcp->super)) + tcp->unorderedInputLength` and the total output buffer length is
`socket_getOutputBufferLength(&(tcp->super)) + tcp->throttledOutputLength`

**Comment by mikeperry-tor** [comment]
@robgjansen - Did you do any tuning of Circuit EWMA in your KIST work? One of our concerns is getting that prioritization to work for us properly, and it is kind of a black box.

**Reply by robgjansen** [reply]
We did some experiments to show in which cases prioritization suffers with vanilla (See Section 5.3 and 5.4 in the journal version), but I'm not finding any experiments where we varied the `CircuitPriorityHalflife` option (we used 30 seconds in our experiments as was the default in the consensus at the time).

Note that many traffic differentiation/prioritization schemes are not well defined when the network utilization rate is below some threshold - they tend to work better under heavier loads. Client load is another thing to experiment with in addition to the EWMA halflife (e.g., by adding extra client downloaders, or removing some -- using the [`--load_scale` option in `tornettools generate`](https://github.com/shadow/tornettools/blob/ac6a9425f6d853510d2cf3a2a9b2cf94b65b1fad/tornettools/tornettools#L299-L308)).
